File: java/src/main/java/com/quant/frtb/Ima.java
package com.quant.frtb;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * Internal Models Approach sketch: ES 97.5% with liquidity-horizon scaling,
 * IMCC, backtesting zones/multipliers, NMRF stress capital (SES).
 *
 * <p>Pinned conventions (see API_SPEC.md):
 * <ul>
 *   <li>ES 97.5 (daily): losses L = -PnL sorted descending;
 *       {@code k = max(1, ceil((1-alpha)*n - 1e-9))} (the epsilon guards
 *       binary-float ceil artefacts); ES = mean of the k worst losses.
 *       Base 10d ES = sqrt(10) * daily ES.</li>
 *   <li>LH ladder (10, 20, 40, 60, 120):
 *       {@code ES_LH = sqrt(ES1(P)^2 + sum_j (ES1(P_j)*sqrt((LH_j-LH_{j-1})/10))^2)}
 *       where P_j sums the categories with horizon &ge; LH_j. Monotone:
 *       ES_LH &ge; base ES.</li>
 *   <li>IMCC = rho*ES_LH(full) + (1-rho)*sum_c ES_LH(category), rho = 0.5.</li>
 *   <li>Backtesting (99% VaR, 260d): exception when PnL_t &lt; -VaR_t; zones
 *       green 0-4, amber 5-9, red &ge; 10; pinned multiplier table.</li>
 *   <li>SES = sum of pinned NMRF stressed losses, zero diversification.</li>
 * </ul>
 */
public final class Ima {

    private Ima() {
    }

    /** One pinned non-modellable risk factor entry from {@code nmrf.json}. */
    public record NmrfEntry(String factor, String desk, double stressedLoss) {
    }

    /** Daily ES at level alpha: mean of the k = ceil((1-alpha)*n) worst losses. */
    public static double expectedShortfallDaily(double[] pnl, double alpha) {
        int n = pnl.length;
        if (n == 0) {
            throw new IllegalArgumentException("expectedShortfallDaily: empty P&L series");
        }
        if (!(0.0 < alpha && alpha < 1.0)) {
            throw new IllegalArgumentException(
                    "expectedShortfallDaily: alpha must be in (0,1), got " + alpha);
        }
        for (double v : pnl) {
            if (!Double.isFinite(v)) {
                throw new IllegalArgumentException(
                        "expectedShortfallDaily: P&L contains non-finite values");
            }
        }
        // tiny epsilon guards against binary-float artefacts like 0.025*40 -> 1.0000...09
        int k = Math.max(1, (int) Math.ceil((1.0 - alpha) * n - 1e-9));
        double[] losses = new double[n];
        for (int i = 0; i < n; i++) {
            losses[i] = -pnl[i];
        }
        Arrays.sort(losses); // ascending; the k worst losses sit at the top
        double sum = 0.0;
        for (int i = n - 1; i >= n - k; i--) {
            sum += losses[i]; // largest first, matching the reference sum order
        }
        return sum / k;
    }

    /** Base 10-day ES: sqrt(10) * daily ES (pinned square-root-of-time scaling). */
    public static double esBase10d(double[] pnl, double alpha) {
        return Math.sqrt(10.0) * expectedShortfallDaily(pnl, alpha);
    }

    /**
     * Liquidity-horizon-scaled ES (Basel ladder formula, see class docs).
     *
     * <p>The category series must sum to the full P&L (validated to 1e-6) and
     * every category must have a pinned liquidity horizon.
     */
    public static double esLhScaled(double[] fullPnl, Map<String, double[]> categoryPnl,
                                    Map<String, Integer> categoryLh, int[] lhLadder,
                                    double alpha) {
        if (lhLadder.length < 1 || !strictlyIncreasing(lhLadder)) {
            throw new IllegalArgumentException("esLhScaled: lhLadder must be strictly increasing");
        }
        if (lhLadder[0] != 10) {
            throw new IllegalArgumentException(
                    "esLhScaled: lhLadder must start at the 10d base horizon");
        }
        int n = fullPnl.length;
        for (Map.Entry<String, double[]> e : categoryPnl.entrySet()) {
            if (!categoryLh.containsKey(e.getKey())) {
                throw new IllegalArgumentException(
                        "esLhScaled: no pinned liquidity horizon for category '"
                                + e.getKey() + "'");
            }
            if (e.getValue().length != n) {
                throw new IllegalArgumentException(
                        "esLhScaled: category '" + e.getKey() + "' length mismatch");
            }
        }
        for (int i = 0; i < n; i++) {
            double s = 0.0;
            for (double[] series : categoryPnl.values()) {
                s += series[i];
            }
            if (Math.abs(s - fullPnl[i]) > 1e-6) {
                throw new IllegalArgumentException(
                        "esLhScaled: category P&L does not sum to the full P&L on day " + i
                                + " (" + s + " vs " + fullPnl[i] + ")");
            }
        }

        double base = esBase10d(fullPnl, alpha);
        double totalSq = base * base;
        for (int j = 1; j < lhLadder.length; j++) {
            int lhJ = lhLadder[j];
            int lhPrev = lhLadder[j - 1];
            double[] subset = new double[n];
            boolean anyCat = false;
            boolean allZero = true;
            for (Map.Entry<String, double[]> e : categoryPnl.entrySet()) {
                if (categoryLh.get(e.getKey()) >= lhJ) {
                    anyCat = true;
                    double[] series = e.getValue();
                    for (int i = 0; i < n; i++) {
                        subset[i] += series[i];
                    }
                }
            }
            if (!anyCat) {
                continue;
            }
            for (double v : subset) {
                if (v != 0.0) {
                    allZero = false;
                    break;
                }
            }
            if (allZero) {
                continue;
            }
            double term = esBase10d(subset, alpha) * Math.sqrt((lhJ - lhPrev) / 10.0);
            totalSq += term * term;
        }
        return Math.sqrt(totalSq);
    }

    private static boolean strictlyIncreasing(int[] xs) {
        for (int i = 1; i < xs.length; i++) {
            if (xs[i] <= xs[i - 1]) {
                return false;
            }
        }
        return true;
    }

    /** IMCC = rho*ES_LH(full) + (1-rho)*sum over categories of ES_LH(category). */
    public static double imcc(double[] fullPnl, Map<String, double[]> categoryPnl,
                              SbmParams params) {
        double rho = params.imaRho();
        double esFull = esLhScaled(fullPnl, categoryPnl, params.categoryLh(),
                params.lhLadder(), params.imaAlpha());
        double esPartials = 0.0;
        for (Map.Entry<String, double[]> e : categoryPnl.entrySet()) {
            esPartials += esLhScaled(e.getValue(), Map.of(e.getKey(), e.getValue()),
                    params.categoryLh(), params.lhLadder(), params.imaAlpha());
        }
        return rho * esFull + (1.0 - rho) * esPartials;
    }

    // ---------------------------------------------------------------------
    // Backtesting
    // ---------------------------------------------------------------------

    /**
     * VaR backtest outcome.
     *
     * @param exceptions number of days with PnL below -VaR (strict)
     * @param zone       "green" | "amber" | "red"
     * @param multiplier pinned capital multiplier for the zone
     */
    public record BacktestResult(int exceptions, String zone, double multiplier) {
    }

    /** Count 99% VaR exceptions (PnL_t &lt; -VaR_t) and map to zone/multiplier. */
    public static BacktestResult backtest(double[] pnl, double[] var99, SbmParams params) {
        if (pnl.length != var99.length) {
            throw new IllegalArgumentException("backtest: P&L and VaR length mismatch ("
                    + pnl.length + " vs " + var99.length + ")");
        }
        if (pnl.length == 0) {
            throw new IllegalArgumentException("backtest: empty series");
        }
        for (double v : var99) {
            if (!Double.isFinite(v) || v < 0.0) {
                throw new IllegalArgumentException(
                        "backtest: VaR values must be non-negative and finite");
            }
        }
        int exceptions = 0;
        for (int i = 0; i < pnl.length; i++) {
            if (pnl[i] < -var99[i]) {
                exceptions++;
            }
        }
        return new BacktestResult(exceptions, backtestZone(exceptions),
                backtestMultiplier(exceptions, params));
    }

    /** Basel traffic-light zone: green 0-4, amber 5-9, red &ge; 10. */
    public static String backtestZone(int exceptions) {
        if (exceptions < 0) {
            throw new IllegalArgumentException("backtestZone: exception count cannot be negative");
        }
        if (exceptions <= 4) {
            return "green";
        }
        if (exceptions <= 9) {
            return "amber";
        }
        return "red";
    }

    /** Pinned multiplier: 1.5 green; amber table 5..9; 2.0 red (cap, also &gt; 12). */
    public static double backtestMultiplier(int exceptions, SbmParams params) {
        String zone = backtestZone(exceptions);
        if ("green".equals(zone)) {
            return params.backtestBaseMultiplier();
        }
        if ("amber".equals(zone)) {
            Double m = params.backtestAmberMultipliers().get(exceptions);
            if (m == null) {
                throw new IllegalArgumentException(
                        "backtestMultiplier: no amber multiplier pinned for " + exceptions);
            }
            return m;
        }
        return params.backtestRedMultiplier();
    }

    // ---------------------------------------------------------------------
    // NMRF / SES
    // ---------------------------------------------------------------------

    /**
     * Stress scenario capital: sum of stressed losses, zero diversification.
     * Negative or non-finite stressed losses raise.
     */
    public static double ses(List<NmrfEntry> entries) {
        double total = 0.0;
        for (NmrfEntry e : entries) {
            double loss = e.stressedLoss();
            if (!Double.isFinite(loss) || loss < 0.0) {
                throw new IllegalArgumentException(
                        "ses: stressed_loss must be >= 0 and finite (factor '"
                                + e.factor() + "')");
            }
            total += loss;
        }
        return total;
    }

    /**
     * IMA capital = multiplier * IMCC + SES + PLAT surcharge.
     *
     * <p>Simplification (documented): avg60(IMCC) = IMCC for the static
     * bundled portfolio, so max(IMCC, m*avg60(IMCC)) = m*IMCC since m &ge; 1.5.
     */
    public static double imaCapital(double imccValue, double multiplier, double sesValue,
                                    double platSurcharge) {
        checkNonNegative("imcc", imccValue);
        checkNonNegative("multiplier", multiplier);
        checkNonNegative("ses", sesValue);
        checkNonNegative("platSurcharge", platSurcharge);
        return multiplier * imccValue + sesValue + platSurcharge;
    }

    private static void checkNonNegative(String name, double v) {
        if (!Double.isFinite(v) || v < 0.0) {
            throw new IllegalArgumentException(
                    "imaCapital: " + name + " must be >= 0 and finite, got " + v);
        }
    }
}

File: python/src/frtb/ima.py
"""Internal Models Approach sketch: ES 97.5% with liquidity-horizon scaling,
IMCC, backtesting zones/multipliers, NMRF stress capital (SES).

Pinned conventions (see API_SPEC.md):

* ES 97.5 (daily): losses L = -PnL sorted descending; k = ceil((1-alpha)*n);
  ES_daily = mean of the k worst losses.  Base 10d ES = sqrt(10) * ES_daily.
* Liquidity-horizon ladder (Basel-style):
      ES_LH = sqrt( ES_1(P)^2 + sum_{j>=2} ( ES_1(P_j) * sqrt((LH_j - LH_{j-1})/10) )^2 )
  where LH ladder = (10, 20, 40, 60, 120), P_j = P&L of the risk-factor
  categories whose pinned liquidity horizon is >= LH_j, and ES_1 is the base
  10d ES operator above.  The ladder is monotone: ES_LH >= base ES.
* IMCC = rho * ES_LH(full) + (1-rho) * sum_c ES_LH(category c), rho = 0.5.
* Capital = max(IMCC_{t-1}, multiplier * avg60(IMCC)); with the bundled static
  portfolio IMCC is constant so avg60(IMCC) = IMCC and the max resolves to
  multiplier * IMCC (multiplier >= 1.5) — documented simplification.
* Backtesting (99% VaR, 260 days): exception when PnL_t < -VaR_t.
  Zones: Green 0-4, Amber 5-9, Red >= 10.  Multiplier: 1.5 (green),
  pinned amber table {5:1.70, 6:1.75, 7:1.83, 8:1.88, 9:1.92}, red 2.0
  (cap — also applies for any count > 12).
* SES = sum of pinned NMRF stressed losses, zero diversification benefit.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Dict, Mapping, Sequence

from .params import SbmParams


def expected_shortfall_daily(pnl: Sequence[float], alpha: float = 0.975) -> float:
    """Daily ES at level alpha: mean of the k = ceil((1-alpha)*n) worst losses."""
    n = len(pnl)
    if n == 0:
        raise ValueError("expected_shortfall_daily: empty P&L series")
    if not (0.0 < alpha < 1.0):
        raise ValueError(f"expected_shortfall_daily: alpha must be in (0,1), got {alpha}")
    for v in pnl:
        if not math.isfinite(v):
            raise ValueError("expected_shortfall_daily: P&L contains non-finite values")
    # tiny epsilon guards against binary-float artefacts like 0.025*40 -> 1.0000000000000009
    k = max(1, math.ceil((1.0 - alpha) * n - 1e-9))
    losses = sorted((-v for v in pnl), reverse=True)
    return sum(losses[:k]) / k


def es_base_10d(pnl: Sequence[float], alpha: float = 0.975) -> float:
    """Base 10-day ES: sqrt(10) * daily ES (pinned square-root-of-time scaling)."""
    return math.sqrt(10.0) * expected_shortfall_daily(pnl, alpha)


def es_lh_scaled(
    full_pnl: Sequence[float],
    category_pnl: Mapping[str, Sequence[float]],
    category_lh: Mapping[str, int],
    lh_ladder: Sequence[int],
    alpha: float = 0.975,
) -> float:
    """Liquidity-horizon-scaled ES (Basel ladder formula, see module docstring).

    ``category_pnl`` series must sum to the full P&L (validated to 1e-6) and
    every category must have a pinned liquidity horizon in ``category_lh``.
    """
    if len(lh_ladder) < 1 or list(lh_ladder) != sorted(set(lh_ladder)):
        raise ValueError("es_lh_scaled: lh_ladder must be strictly increasing")
    if lh_ladder[0] != 10:
        raise ValueError("es_lh_scaled: lh_ladder must start at the 10d base horizon")
    n = len(full_pnl)
    for cat, series in category_pnl.items():
        if cat not in category_lh:
            raise ValueError(f"es_lh_scaled: no pinned liquidity horizon for category '{cat}'")
        if len(series) != n:
            raise ValueError(f"es_lh_scaled: category '{cat}' length mismatch")
    for i in range(n):
        s = sum(category_pnl[c][i] for c in category_pnl)
        if abs(s - full_pnl[i]) > 1e-6:
            raise ValueError("es_lh_scaled: category P&L does not sum to the full P&L "
                             f"on day {i} ({s} vs {full_pnl[i]})")

    total_sq = es_base_10d(full_pnl, alpha) ** 2
    for j in range(1, len(lh_ladder)):
        lh_j, lh_prev = lh_ladder[j], lh_ladder[j - 1]
        cats = [c for c in category_pnl if category_lh[c] >= lh_j]
        if not cats:
            continue
        subset = [sum(category_pnl[c][i] for c in cats) for i in range(n)]
        if all(v == 0.0 for v in subset):
            continue
        term = es_base_10d(subset, alpha) * math.sqrt((lh_j - lh_prev) / 10.0)
        total_sq += term * term
    return math.sqrt(total_sq)


def imcc(
    full_pnl: Sequence[float],
    category_pnl: Mapping[str, Sequence[float]],
    params: SbmParams,
) -> float:
    """IMCC = rho * ES_LH(full) + (1-rho) * sum over categories of ES_LH(category)."""
    rho = params.ima_rho
    es_full = es_lh_scaled(full_pnl, category_pnl, params.category_lh,
                           params.lh_ladder, params.ima_alpha)
    es_partials = 0.0
    for cat, series in category_pnl.items():
        es_partials += es_lh_scaled(series, {cat: series}, params.category_lh,
                                    params.lh_ladder, params.ima_alpha)
    return rho * es_full + (1.0 - rho) * es_partials


# --------------------------------------------------------------------------
# Backtesting
# --------------------------------------------------------------------------

@dataclass(frozen=True)
class BacktestResult:
    """VaR backtest outcome: exception count, Basel zone, capital multiplier."""

    exceptions: int
    zone: str  # 'green' | 'amber' | 'red'
    multiplier: float


def backtest(pnl: Sequence[float], var99: Sequence[float], params: SbmParams) -> BacktestResult:
    """Count 99% VaR exceptions (PnL_t < -VaR_t) and map to zone/multiplier."""
    if len(pnl) != len(var99):
        raise ValueError(f"backtest: P&L and VaR length mismatch ({len(pnl)} vs {len(var99)})")
    if len(pnl) == 0:
        raise ValueError("backtest: empty series")
    for v in var99:
        if not math.isfinite(v) or v < 0.0:
            raise ValueError("backtest: VaR values must be non-negative and finite")
    exceptions = sum(1 for p, v in zip(pnl, var99) if p < -v)
    return BacktestResult(exceptions, backtest_zone(exceptions),
                          backtest_multiplier(exceptions, params))


def backtest_zone(exceptions: int) -> str:
    """Basel traffic-light zone: green 0-4, amber 5-9, red >= 10."""
    if exceptions < 0:
        raise ValueError("backtest_zone: exception count cannot be negative")
    if exceptions <= 4:
        return "green"
    if exceptions <= 9:
        return "amber"
    return "red"


def backtest_multiplier(exceptions: int, params: SbmParams) -> float:
    """Pinned multiplier: 1.5 green; amber table 5..9; 2.0 red (cap, also > 12)."""
    zone = backtest_zone(exceptions)
    if zone == "green":
        return params.backtest_base_multiplier
    if zone == "amber":
        if exceptions not in params.backtest_amber_multipliers:
            raise ValueError(f"backtest_multiplier: no amber multiplier pinned for {exceptions}")
        return params.backtest_amber_multipliers[exceptions]
    return params.backtest_red_multiplier


# --------------------------------------------------------------------------
# NMRF / SES
# --------------------------------------------------------------------------

def ses(nmrf_entries: Sequence[Mapping[str, object]]) -> float:
    """Stress scenario capital: sum of stressed losses, zero diversification.

    Each entry: {"factor": str, "desk": str, "stressed_loss": float >= 0}.
    """
    total = 0.0
    for e in nmrf_entries:
        loss = float(e["stressed_loss"])  # type: ignore[arg-type]
        if not math.isfinite(loss) or loss < 0.0:
            raise ValueError(f"ses: stressed_loss must be >= 0 and finite "
                             f"(factor '{e.get('factor')}')")
        total += loss
    return total


def ima_capital(imcc_value: float, multiplier: float, ses_value: float,
                plat_surcharge: float = 0.0) -> float:
    """IMA capital = multiplier * IMCC + SES + PLAT surcharge.

    Simplification (documented): avg60(IMCC) = IMCC for the static bundled
    portfolio, so max(IMCC, m*avg60(IMCC)) = m*IMCC since m >= 1.5.
    """
    for name, v in (("imcc", imcc_value), ("multiplier", multiplier),
                    ("ses", ses_value), ("plat_surcharge", plat_surcharge)):
        if not math.isfinite(v) or v < 0.0:
            raise ValueError(f"ima_capital: {name} must be >= 0 and finite, got {v}")
    return multiplier * imcc_value + ses_value + plat_surcharge
